File: Auth/Users.App/ChangePasswordService.cs
using System.Text.RegularExpressions;
using Users.App.Interface;

namespace Users.App
{
    public class ChangePasswordService(IUserRepo repo)
    {
        public async Task<ChangePasswordResult> ChangePasswordAsync(ChangePasswordRequest request)
        {
            var user = await repo.FindByIdAsync(request.UserId);
            if (user == null)
            {
                return new ChangePasswordResult
                {
                    Success = false,
                    Error = "UserNotFound",
                    ErrorDescription = "User was not found."
                };
            }

            if (request.LockEnabled)
            {
                if (user.IsLocked || (user.LockedUntil.HasValue && DateTimeOffset.UtcNow <= user.LockedUntil.Value))
                {
                    return new ChangePasswordResult
                    {
                        Success = false,
                        Error = "AccountLocked",
                        ErrorDescription = "Account is locked."
                    };
                }
            }

            if (Constants.IsLocalAccount(user.Issuer) == false)
            {
                return new ChangePasswordResult
                {
                    Success = false,
                    Error = "InvalidAccountType",
                    ErrorDescription = "Cannot change password for federated accounts."
                };
            }

            if (user.Password == null || !PasswordHasher.Verify(user.Password, request.OldPassword))
            {
                return new ChangePasswordResult
                {
                    Success = false,
                    Error = "InvalidOldPassword",
                    ErrorDescription = "Old password is incorrect."
                };
            }

            user.Password = PasswordHasher.Hash(request.NewPassword);
            await repo.UpdateAsync(user);

            return new ChangePasswordResult { Success = true };
        }
    }

    public class ChangePasswordResult
    {
        public bool Success { get; set; }
        public string? Error { get; set; }
        public string? ErrorDescription { get; set; }
    }
}

File: NodeBackApi/src/services/ChangePasswordService.ts
import { isLocalAccount } from "../utils/constants";
import { PasswordHasher } from "../utils/passwordHasher";
import type { IUserRepo } from "../repos/interfaces/IUserRepo";

export interface ChangePasswordRequest {
  userId: string;
  oldPassword: string;
  newPassword: string;
  lockEnabled: boolean;
}

export interface ChangePasswordResult {
  success: boolean;
  error?: string;
  errorDescription?: string;
}

export class ChangePasswordService {
  constructor(private readonly repo: IUserRepo) {}

  async changePasswordAsync(
    request: ChangePasswordRequest,
  ): Promise<ChangePasswordResult> {
    const user = await this.repo.findByIdAsync(request.userId);
    if (!user) {
      return {
        success: false,
        error: "UserNotFound",
        errorDescription: "User was not found.",
      };
    }

    if (request.lockEnabled) {
      if (
        user.isLocked ||
        (user.lockedUntil && Date.now() <= user.lockedUntil.getTime())
      ) {
        return {
          success: false,
          error: "AccountLocked",
          errorDescription: "Account is locked.",
        };
      }
    }

    if (!isLocalAccount(user.issuer)) {
      return {
        success: false,
        error: "InvalidAccountType",
        errorDescription: "Cannot change password for federated accounts.",
      };
    }

    if (!user.password || !PasswordHasher.verify(user.password, request.oldPassword)) {
      return {
        success: false,
        error: "InvalidOldPassword",
        errorDescription: "Old password is incorrect.",
      };
    }

    user.password = PasswordHasher.hash(request.newPassword);
    await this.repo.updateAsync(user);

    return { success: true };
  }
}
